core/src/main/kotlin/fr/sncf/osrd/pathfinding/constraints/TrackSectionConstraints.kt

```diff
@@ -0,0 +1,45 @@
+package fr.sncf.osrd.pathfinding.constraints
+
+import fr.sncf.osrd.graph.PathfindingConstraint
+import fr.sncf.osrd.path.implementations.buildTrainPathFromBlock
+import fr.sncf.osrd.path.interfaces.TrainPath
+import fr.sncf.osrd.pathfinding.Pathfinding
+import fr.sncf.osrd.sim_infra.api.Block
+import fr.sncf.osrd.sim_infra.api.BlockId
+import fr.sncf.osrd.sim_infra.api.BlockInfra
+import fr.sncf.osrd.sim_infra.api.RawSignalingInfra
+import fr.sncf.osrd.sim_infra.api.TrackSectionId
+
+data class TrackSectionConstraints(
+    val blockInfra: BlockInfra,
+    val infra: RawSignalingInfra,
+    val trackSectionIds: Set<TrackSectionId>?,
+) : PathfindingConstraint<Block> {
+    override fun apply(edge: BlockId): Collection<Pathfinding.Range<Block>> {
+        if (trackSectionIds == null) return setOf()
+
+        val res = HashSet<Pathfinding.Range<Block>>()
+        val path = buildTrainPathFromBlock(infra, blockInfra, edge)
+        res.addAll(getBlockedRanges(trackSectionIds, path))
+
+        return res
+    }
+
+    private fun getBlockedRanges(
+        trackSectionIds: Set<TrackSectionId>,
+        path: TrainPath,
+    ): Collection<Pathfinding.Range<Block>> {
+        val res = HashSet<Pathfinding.Range<Block>>()
+
+        val invalidTrackChunks =
+            path.getChunks().filter {
+                val trackSectionId = infra.getTrackFromChunk(it.value.value)
+                !trackSectionIds.contains(trackSectionId)
+            }
+
+        for (range in invalidTrackChunks) {
+            res.add(Pathfinding.Range(range.pathBegin.cast(), range.pathEnd.cast()))
```

Comment:
The cast is better, now it's just missing a comment

```suggestion
            // The train path contains exactly the given block,
            // so we can cast offsets from one type to the other
            res.add(Pathfinding.Range(range.pathBegin.cast(), range.pathEnd.cast()))
```